### PR TITLE
Added support for MatrixPlot for use in Heatmaps

### PR DESCRIPTION
**Pull request by AnkurD**
Referencing issue #73 
I borrowed a lot of implementation from the Image command, adjusting it to allow for MatrixPlot to be used and for NaN to be supported. The render time for large matrices is far too long to be tenable, so I set up a fallback to use the Image method for rendering larger matrices as PNGs instead.

### CODE REVIEW

File: src/colormaps.jl [hunk 1]
```diff
@@ -86,6 +86,7 @@ end
 function Base.write(colormap::RGBArrayMap, data, filename)
     colors = interpolate_RGBArrayMap(colormap)
     n = length(colors)
+	data = clamp01nan.(data)
```
Comment (rejuvyesh) [inline src/colormaps.jl]: Indentation issue!

File: src/plots.jl [hunk 2]
```diff
@@ -418,4 +419,96 @@ function Histogram2(
...
 
+mutable struct MatrixPlot <: Plot
+    filename::AbstractString
+    xmin::Real
+    xmax::Real
+    ymin::Real
+    ymax::Real
+    zmin::Real
+    zmax::Real
+	rows::Real
```
Comment (rejuvyesh) [inline src/plots.jl]: Indentation

File: src/plots.jl [hunk 3]
```diff
@@ -418,4 +419,96 @@ function Histogram2(
...
+	cols::Real
+    zmode
+	raster
```
Comment (rejuvyesh) [inline src/plots.jl]: Indentation

File: src/PGFPlots.jl [hunk 4]
```diff
@@ -656,6 +656,25 @@ function plotHelper(o::IO, p::Patch2D)
     plotLegend(o, p.legendentry)
 end
 
+function plotHelper(o::IO, p::MatrixPlot)
+    if p.zmin == p.zmax
+        error("Your colorbar range limits must not be equal to each other.")
```
Comment (tawheeler) [inline src/PGFPlots.jl]: Should this be `p.zmin >= p.zmax`?

File: src/PGFPlots.jl [hunk 5]
```diff
@@ -787,6 +807,19 @@ function axisOptions(p::Image)
     end
 end
 
+function axisOptions(p::MatrixPlot)
+    if p.colorbar
+        cmOpt = colormapOptions(p.colormap)
+        if p.colorbarStyle == nothing
+            return "enlargelimits = false, axis on top, $cmOpt, colorbar,xmin=$(p.xmin), xmax=$(p.xmax), ymin=$(p.ymin), ymax=$(p.ymax)"
+        else
+            return "enlargelimits = false, axis on top, $cmOpt, colorbar, colorbar,xmin=$(p.xmin), xmax=$(p.xmax), ymin=$(p.ymin), ymax=$(p.ymax), style = {$(p.colorbarStyle)}"
```
Comment (tawheeler) [inline src/PGFPlots.jl]: repeated colorbar

File: src/PGFPlots.jl [hunk 6]
```diff
@@ -787,6 +807,19 @@ function axisOptions(p::Image)
     end
 end
 
+function axisOptions(p::MatrixPlot)
+    if p.colorbar
+        cmOpt = colormapOptions(p.colormap)
+        if p.colorbarStyle == nothing
+            return "enlargelimits = false, axis on top, $cmOpt, colorbar,xmin=$(p.xmin), xmax=$(p.xmax), ymin=$(p.ymin), ymax=$(p.ymax)"
```
Comment (tawheeler) [inline src/PGFPlots.jl]: spacing after colorbar

File: src/PGFPlots.jl [hunk 7]
```diff
@@ -787,6 +807,19 @@ function axisOptions(p::Image)
     end
 end
 
+function axisOptions(p::MatrixPlot)
+    if p.colorbar
+        cmOpt = colormapOptions(p.colormap)
+        if p.colorbarStyle == nothing
```
Comment (tawheeler) [inline src/PGFPlots.jl]: It'd be easier to understand if this was changed to:
```julia
option_string = "enlargelimits = false, axis on top, $cmOpt, colorbar, xmin=$(p.xmin), xmax=$(p.xmax), ymin=$(p.ymin), ymax=$(p.ymax)"
if p.colorbarStyle == nothing
    option_string *= ", style = {$(p.colorbarStyle)}"
end
return option_string

File: src/plots.jl [hunk 8]
```diff
@@ -418,4 +419,96 @@ function Histogram2(
...
+			yrange = (1,rows)
+		end
+        if zmin == nothing
+            zmin = minimum(A[.!isnan.(A)])
+        end
+        if zmax == nothing
+            zmax = maximum(A[.!isnan.(A)])
+        end
+        if zmin == zmax
+            zmin -= 1.
```
Comment (tawheeler) [inline src/plots.jl]: Can '0's be added to these floating point values? It makes it more explicit.

File: src/plots.jl [hunk 9]
```diff
@@ -418,4 +419,96 @@ function Histogram2(
...
+		elseif raster == nothing
+			@warn "Matrix is too large for vector plotting, consider switching to raster mode."
+			raster = true			
+		end
+		
+		if raster
+			file, _ = splitext(filename)
+			filename = string(file,".png")
+			A = clamp.(A, zmin, zmax)
+			A = A .- zmin
```
Comment (tawheeler) [inline src/plots.jl]:
```
A .-= zmin
A ./= (zmax - zmin)
```

or 

```
A[:] = (A .- zmin) ./ (zmax - zmin)
```

File: src/plots.jl [hunk 10]
```diff
@@ -418,4 +419,96 @@ function Histogram2(
...
+			if isa(colormap, ColorMaps.ColorMap)
+				write(colormap, A, filename)
+			else
+				write(ColorMaps.RGBArrayMap(colormap), A, filename)
+			end
+		else
+			out = Array{Any}(undef,length(A)+1,3)
+			x = range(xrange[1], stop=xrange[2], length=cols)
+			y = range(yrange[1], stop=yrange[2], length=rows)
+			x = x .- 0.5*step(x)
```
Comment (tawheeler) [inline src/plots.jl]: Slight preference for
```
x .-= step(x)/2
y .-= step(y)/2
```

File: src/PGFPlots.jl (hunk 7 above)
Comment (AnkurD) [reply src/PGFPlots.jl]: I think it needs to be the other way around, so I adjusted it to this
```julia
option_string = "enlargelimits = false, axis on top, $cmOpt, colorbar, xmin=$(p.xmin), xmax=$(p.xmax), ymin=$(p.ymin), ymax=$(p.ymax)"
if p.colorbarStyle != nothing
    option_string *= ", style = {$(p.colorbarStyle)}"
end
    return option_string
```

File: src/plots.jl (hunk 10 above)
Comment (AnkurD) [reply src/plots.jl]: So I tried this initially, but I get an error
```julia
ERROR: setindex! not defined for StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}```.
For the time being I'll leave it as is, but if there's something I missed in the implementation I can go back and fix it.

**Comment by AnkurD** [comment]
Thank you both for the comments! Let me know if there's anything else I should fix.

**Comment by mykelk** [comment]
Do the tests pass on your local machine? It looks like there is a failure on Travis related to testing the notebook.

**Comment by mykelk** [comment]
Thank you for your contribution!

**Comment by mykelk** [comment]
Oh, would you mind reevaluating all the cells in the documentation notebook so that your plots show up properly? Right now the output of those cells are now showing up.

**Pull request merged**